lib/chef/provisioning/managed_entry.rb

```diff
@@ -0,0 +1,121 @@
+class Chef
+module Provisioning
+  #
+  # Specification for a managed thing.  Remembers where it was stored, and lets
+  # you stuff reference data in it.
+  #
+  class ManagedEntry
+    def initialize(managed_entry_store, resource_type, name, data=nil)
+      @managed_entry_store = managed_entry_store
+      @resource_type = resource_type
+      @name = name
+      @data = data || {}
+    end
+
+    attr_reader :managed_entry_store
+    attr_reader :resource_type
+    attr_reader :name
+    attr_reader :data
+
+    def attrs
+      data
+    end
+
+    #
+    # Globally unique identifier for this machine. Does not depend on the machine's
+    # reference or existence.
+    #
+    def id
+      managed_entry_store.identifier(resource_type, name)
+    end
+
+    #
+    # Reference to this managed thing.  This should be a freeform hash, with enough
+    # information for the driver to look it up and create a Machine object to
+    # access it.
+    #
+    # This MUST include a 'driver_url' attribute with the driver's URL in it.
+    #
+    # chef-provisioning will do its darnedest to not lose this information.
+    #
+    def reference
+      # Backcompat: old data bags didn't have the "reference" field.  If we have
+      # no reference field in the data, and the data bag is non-empty, return
+      # the root of the data bag.
+      attrs['reference'] || attrs['location'] || (attrs == {} ? nil : attrs)
```

Comment:
Same `nil` value question as above. Again, it may not be an issue, it depends on the potential and actual values allowed in `attrs`.